File: php/src/LombokAlgoritma/Hash/Fnv1a.php
<?php

// LombokAlgoritma — FNV-1a 32/64-bit (SPEC §12)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok

declare(strict_types=1);

namespace LombokAlgoritma\Hash;

use LombokAlgoritma\Core\U32;
use LombokAlgoritma\Core\U64;

/** Fowler–Noll–Vo FNV-1a over bytes (a PHP string; UTF-8 for text). Non-cryptographic. */
final class Fnv1a
{
    private const OFFSET64 = -3750763034362895579; // 0xcbf29ce484222325
    private const PRIME64 = 0x100000001b3;

    /** FNV-1a 32-bit as an unsigned int. */
    public static function hash32(string $data): int
    {
        $h = 0x811c9dc5;
        $n = strlen($data);
        for ($i = 0; $i < $n; $i++) {
            $h = U32::mul($h ^ ord($data[$i]), 0x01000193);
        }
        return $h;
    }

    /** FNV-1a 64-bit as a u64 bit pattern (PHP int). */
    public static function hash64Int(string $data): int
    {
        $h = self::OFFSET64;
        $n = strlen($data);
        for ($i = 0; $i < $n; $i++) {
            $h = U64::mul($h ^ ord($data[$i]), self::PRIME64);
        }
        return $h;
    }

    /** FNV-1a 64-bit as 16 lowercase hex digits. */
    public static function hash64(string $data): string
    {
        return U64::toHex(self::hash64Int($data));
    }
}

File: rust/lombokalgoritma/src/string/hash/fnv1a.rs
// LombokAlgoritma — FNV-1a (SPEC §12)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok

/// FNV-1a, 32-bit (offset `0x811c9dc5`, prime `0x01000193`).
pub fn fnv1a32(data: &[u8]) -> u32 {
    let mut h: u32 = 0x811c_9dc5;
    for &b in data {
        h ^= u32::from(b);
        h = h.wrapping_mul(0x0100_0193);
    }
    h
}

/// FNV-1a, 64-bit (offset `0xcbf29ce484222325`, prime `0x100000001b3`).
pub fn fnv1a64(data: &[u8]) -> u64 {
    let mut h: u64 = 0xcbf2_9ce4_8422_2325;
    for &b in data {
        h ^= u64::from(b);
        h = h.wrapping_mul(0x0000_0100_0000_01b3);
    }
    h
}
